### PR TITLE
Display Jupyter/IPython notebooks

### PR DESCRIPTION
**Pull request by hroncok**
https://github.com/pyvec/naucse.python.cz/issues/50

Needs:

 * [x] CSS
 * [x] LaTeX
 * [x] Maybe tests

### CODE REVIEW

**Comment by hroncok** [comment]
Now about LaTeX. Notebook goes with MathJax. As @encukou said, this could end up in a book. So MathJax is not an option? If we want to make LaTeX work, we need to extend `convert_markdown()` to support it (as the LaTeX here is always just in the markdown cells).

**Comment by hroncok** [comment]
Images are base64 encoded and included inline in the HTML. I believe this is not worthy to change (I think we would need to create routes dynamically).

**Comment by hroncok** [comment]
[KaTeX](https://khan.github.io/KaTeX/) is JavaScript as well, but offers server side rendering. I don't know any Python server side LaTeX renderer.

**Comment by encukou** [comment]
I'd say it's something to solve when the book comes along, then. Use JS, but open an issue for it.

It would be nice to track how widely math is used. Maybe put a `latex_math` field in the YAML & model, and refuse to render math without it?

**Comment by hroncok** [comment]
> It would be nice to track how widely math is used. Maybe put a `latex_math` field in the YAML & model, and refuse to render math without it?

Could do.


I'd still rather use KaTeX over MathJax, as it looks better. Thoughts?

**Comment by hroncok** [comment]
I've enabled LaTeX JS conversion everywhere.

In order to only do it in enabled lessons, I could:

 * only link the CSS, load the JS and call the JS on such pages (not sure how)
 * link and load the JS everywhere, but call it only on such pages
   * (by appending the call to the converted HTML maybe, but that would put it before the JS loading, so maybe the same technique as above would ahve to be used anyway)
 * link, load and call the JS everywhere, but restrict it to an element with specific id (such as `latex_container`)  and only add the ID when enabled

I think the last part is the easiest one to implement and makes the CSS and JS being cached sooner than needed - thus makes the page rendering with math faster if the visitor came trough different pages.

**Comment by encukou** [comment]
Pages can currently [define extra css](https://github.com/pyvec/naucse.python.cz/blob/master/lessons/git/basics/info.yml). I think it's a hack for inline CSS, but using something similar for extra CSS/JS files would be useful.

See [models.py](https://github.com/pyvec/naucse.python.cz/blob/master/naucse/models.py#L50) and [templates/lesson.html](https://github.com/pyvec/naucse.python.cz/blob/master/naucse/templates/lesson.html#L3)

**Comment by hroncok** [comment]
Oh, I haven't noticed I have page object in the template. Than that's easy.

**Comment by hroncok** [comment]
Rebased.

File: requirements.txt [hunk 1]
```diff
@@ -1,9 +1,13 @@
 PyYAML
 ansi2html
+-e git+https://github.com/hroncok/elsa.git@frozen_flask_warnings#egg=elsa
```
Comment (encukou) [inline requirements.txt]: I just got around to the review there. Are you planning to release Elsa 0.1.3 for this?

File: lessons/intro/numpy/index.ipynb [hunk 2]
```diff
@@ -3949,20 +3949,6 @@
...
-   "source": [
-    "# Úkol"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "Zadání úkolu je na Eduxu."
-   ]
```
Comment (encukou) [inline lessons/intro/numpy/index.ipynb]: This works. Just a heads up: I'm planning to add actual links to the homework as I go through the course.

File: naucse/notebook_util.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+from nbconvert.filters.highlight import Highlight2HTML, _pygments_highlight
+import nbformat
+import traitlets
+import pygments
+
+from . markdown_util import convert_markdown
+
+
+class Highlight2HTMLCodehilite(Highlight2HTML):
+    '''We need to do all this just to add the codehilite cssclass'''
```
Comment (encukou) [inline naucse/notebook_util.py]: Using `_pygments_highlight` sounds like asking for trouble.

Can we switch to using the same highlighter code as the rest of naucse, or does `_pygments_highlight` do something special?

Alternately, we don't need to use `codehilite` for the class. It was a default from the old Markdown parser. We could just switch to `highlight` if that's what Notebook uses by default.
I can elaborate, if that sounds like a good idea.

File: naucse/static/css/ipython.css [hunk 4]
```diff
@@ -0,0 +1,59 @@
+/*!
+ * Bootstrap v3.3.7 (http://getbootstrap.com)
+ * Copyright 2011-2016 Twitter, Inc.
+ * Licensed under MIT (https://github.com/twbs/bootstrap/blob/master/LICENSE)
+ */
```
Comment (encukou) [inline naucse/static/css/ipython.css]: Um, is this really Bootstrap code? Where does this come from?

File: naucse/static/css/ipython.css [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+.prompt {
+  /* This needs to be wide enough for 3 digit prompt numbers: In[100]: */
+  min-width: 14ex;
+  /* This padding is tuned to match the padding on the CodeMirror editor. */
+  padding: 0.4em;
+  margin: 0px;
+  font-family: monospace;
+  text-align: right;
+  /* This has to match that of the the CodeMirror class line-height below */
```
Comment (encukou) [inline naucse/static/css/ipython.css]: Some context is missing.

File: naucse/notebook_util.py (hunk 3 above)
Comment (encukou) [inline naucse/notebook_util.py]: Also, as things are in this PR, highlighting doesn't work properly for IPython extensions. See the `! python -m pip install` command in the Pandas lesson.

Comment (hroncok) [inline naucse/notebook_util.py]: Will look at this. Probably today.

File: naucse/static/css/ipython.css (hunk 4 above)
Comment (hroncok) [inline naucse/static/css/ipython.css]: From  [nbconvert](https://github.com/jupyter/nbconvert/blob/5.1.1/setup.py#L70) -> [notebook](https://cdn.jupyter.org/notebook/4.3.0/style/style.min.css)

File: requirements.txt (hunk 1 above)
Comment (hroncok) [inline requirements.txt]: Yes.

Comment (hroncok) [inline requirements.txt]: Done. Amended.

File: naucse/static/css/ipython.css (hunk 5 above)
Comment (hroncok) [inline naucse/static/css/ipython.css]: I can remove that comment, it's a copy paste leftover.

**Comment by hroncok** [comment]
> Also, as things are in this PR, highlighting doesn't work properly for IPython extensions. See the `! python -m pip install` command in the Pandas lesson.

pandas_pip_install

Seems OK to me. What was problematic here?

I'v changed `codehilite` to `highlight` and got rid of the redundant code.

**Comment by hroncok** [comment]
Rebased on master.

**Comment by encukou** [comment]
Looks great, thanks!

**Pull request merged**